### DISCUSSION TITLE
Help Request: ./gradlew serverRun leads to PSQLException

### DISCUSSION THREAD

**Discussion by NeoHW** [TEAMMATES/teammates#13139] [Help requests] [answered]
### Environment

- **Commit SHA:** 
- **Operating System:** MacOS Sonama 14.5 
- **JDK Version:** 11
- **Node.js Version:**
- **Firefox / Chrome Version (if applicable):**

### Intention

When doing `./gradlew serverRun`, I get the logs as attached. 
 
Logs: https://drive.google.com/file/d/1KrQhu9FtSjS_XU1hikxgnGSBbWlEY1gj/view?usp=sharing

My guess from the logs is that this is causing the issue:
`org.postgresql.util.PSQLException: FATAL: role "teammates" does not exist`
 
But, i am able to use docker desktop exec and run psql -U teammates successfully.
 
Some other information is that i am able to access http://localhost:8080/web/front/home, but not able to login as user or adminstrator.
 
Localhost:8080 shows  HTTP ERROR 503 Service Unavailable. 

Some screenshots are attached:
* Am able to access http://localhost:8080/web/front/home but when trying http://localhost:8080/web/admin/home, gets an `error not authorised`
![Screenshot 2024-06-30 at 1 17 43 AM](https://github.com/TEAMMATES/teammates/assets/94470439/79a18c82-1b2e-4398-97d2-87b46d43aa2d)

* `503 Service Unavailable` when student/instructor login redirects to localhost:8080
![Screenshot 2024-06-30 at 1 17 58 AM](https://github.com/TEAMMATES/teammates/assets/94470439/ea25a79c-9574-4c01-b2b7-2b1740e1e4b6)

 
Any help would be greatly appreciated!

### Resolution Attempt

* `docker ps` shows that all images are up and running
* Able to use exec and log in as teammates on postgres container
* Recloned repo, re-setup dev environment, but still running into same issue.
* Check to see that teammates db exists
![Screenshot 2024-06-30 at 2 29 58 AM](https://github.com/TEAMMATES/teammates/assets/94470439/30c8aa29-45de-45d5-9ed0-e72d308f6480)

Additional Information:
Was able to somehow get it up and working in the past once randomly, but after restarting the laptop, am facing this issue again.

when doing `docker compose up -d`, datastore might not be working correctly due to image/platform issues?
```
> docker-compose up -d
WARN[0000] /Users/haowei/codes/teammates/docker-compose.yml: `version` is obsolete 
[+] Running 5/5
 ✔ Network teammates_default                                                                                                                                Created                                      0.0s 
 ✔ Container teammates-postgres-1                                                                                                                           Started                                      1.0s 
 ✔ Container teammates-solr-1                                                                                                                               Started                                      1.0s 
 ✔ Container teammates-datastore-1                                                                                                                          Started                                      1.0s 
 ! datastore The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested  
 ```

**Comment by damithc** [comment]
@xenosf @domoberzin @mingyuanc @dishenggg any help?

**Comment by dishenggg** [comment]
Maybe you can try some of these if you havent already.

Check that `teammates/src/main/resources/build-dev.properties` has correct postgres config

Try doing this:
1. `docker compose down` 
2. deleting `teammates/postgres-data` 
3. `docker compose up -d`

If that does not work you can try connecting to the db from outside of the docker container at `localhost:5432` using a db client(Dbeaver, Beekeeper etc.) of your choice and check if the role is there. 
![image](https://github.com/TEAMMATES/teammates/assets/21008450/010956ae-10d1-4d0b-bcae-2208ab8c8838)


Else maybe try something from [here ](https://stackoverflow.com/questions/60193781/postgres-with-docker-compose-gives-fatal-role-root-does-not-exist-error) or [here](https://stackoverflow.com/questions/68626017/docker-compose-psql-error-fatal-role-postgres-does-not-exist)

**Reply by NeoHW** [reply]
Hi @dishenggg 

deleting `teammates/postgres-data` does not work

I have tried connecting to the db using Dbeaver, but the same error message appears : `FATAL: role "teammates" does not exist`
![Screenshot 2024-07-03 at 9 50 02 PM](https://github.com/TEAMMATES/teammates/assets/94470439/73c71847-2519-4604-b71b-525b6d46a1d0)
![Screenshot 2024-07-03 at 9 52 24 PM](https://github.com/TEAMMATES/teammates/assets/94470439/1b591b2f-9309-4ea9-b49d-e897ef9b3602)

I have set Database name and Username to `teammates` in accordance to the postgres env in the `docker-compose.yml` file.
Would like to confirm if that's indeed correct?
Also to check, in the postgres db, is it configured such that there will only be a teammates user, and no default postgres user?

I have tried the two stack overflow links too and also other solutions from past discussions but to no avail.

**Reply by dishenggg** [reply]
Hmmm you have the same settings as me in Dbeaver and there is no default postgres user for me as well, maybe someone with a mac might be able to help you

Also maybe you can check if the port in docker is correct, it should show `5432:5432`

**Comment by mingyuanc** [comment]
Hey, did you try connecting with another tool? Or maybe trying to connect using a connection URI. If those issues still persist could you try changing username to something else.

**Comment by NeoHW** [comment, marked as answer]
Hi, thanks for all the help. After some tinkering, I discovered the error and wanted to share it here in case anyone else encounters the same issue in the future.

The problem arises when a PostgreSQL service is already started using Homebrew and then a PostgreSQL container is started using Docker Compose. The program defaults to using the local PostgreSQL database from Homebrew instead of the Dockerized instance.

To fix this, create the necessary role and database in the local PostgreSQL instance to keep it running.
